---
title: "ATAC-seq MEIRLOP benchmark"
author: "Leonor Schubert Santana"
date: "`r Sys.Date()`"
site: workflowr::wflow_site
output:
   workflowr::wflow_html:
    toc: false
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=T, warning=F, message=F)
```

# Introduction

Underneath are some summary plots shown for the output of MEIRLOP for all data sets in the benchmark. MEIRLOP was run for each experiment in each data set using background peaks calculated by a differential peak analysis. For the plots underneath we calculated for each experiment on which rank MEIRLOP finds the transcription factor examined in the study and then plot the distribution of all the ranks grouped by different variables of interest.

```{r data_preparation}
# Load libraries
library(tidyverse)
library(cowplot)
library(ggsci)

# Load utils.functions file
source("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/code/R/utils_functions.R")

# Define output path
out.path <- '/da/dmp/cb/steinse5/projects/tf_tool_benchmark/output/benchmark_data/atacseq/meirlop/'

# Read in all knownResults.txt output files from meirlop for every dataset
meirlop.txts <- list.files("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/data/atacseq/",
                          recursive = T, full.names = T, pattern = 'lr_results.tsv')

# Create dataframe with dataset, experiment and motif name and their p-value
readInMeirlopTxt <- function(meirlop.txt) {
  conds <- unlist(str_split(meirlop.txt, pattern = '/'))
  meirlop.results <- read_tsv(meirlop.txt, show_col_types = FALSE) %>%
    dplyr::rename(motif_name = 1, padj = padj) %>%
    mutate(study = conds[11], label = conds[16], possible_ranks=length(motif_name),
           motif_name=sub(" .*", "", motif_name)) %>%
    dplyr::select(study, label, motif_name, padj, possible_ranks)
 
  meirlop.results
} 

# Check if ranking results *.rds  exists
if(!file.exists(file.path(out.path, "MEIRLOP_benchmark.RDS"))) {
  # Apply both function to the all experiments
  meirlop.results <- lapply(meirlop.txts, readInMeirlopTxt)  %>% bind_rows
  
  meirlop.list <- list('using no background peaks' = meirlop.results)
  
  # Create empty topTen list and ROC/PR list for unstringent and stringent and info
  top.ten <- sapply(names(meirlop.list),function(x) NULL)
  roc.pr.str <- sapply(names(meirlop.list),function(x) NULL)
  roc.pr.unstr <- sapply(names(meirlop.list),function(x) NULL)
  info.df <- sapply(names(meirlop.list),function(x) NULL)
  topMotifs.df <- sapply(names(meirlop.list),function(x) NULL)
  
  
  # Create rankings 
  for(meirlop.name in names(meirlop.list)) {
    l <- createSummaryDF(meirlop.list[[meirlop.name]], "padj", desc=FALSE, 
                         lambert=TRUE)
    
    meirlop.list[[meirlop.name]] <- l$df
    top.ten[[meirlop.name]] <- l$top.ten
    roc.pr.str[[meirlop.name]] <- l$roc.pr.str
    roc.pr.unstr[[meirlop.name]] <- l$roc.pr.unstr
    info.df[[meirlop.name]] <- l$info.df
    topMotifs.df[[meirlop.name]] <- l$topMotifs.df
  }
} else {
  # Read results
  meirlop.list <- read_rds(file.path(out.path, "MEIRLOP_benchmark.RDS"))
  top.ten <- read_rds(file.path(out.path, "MEIRLOP_topTen.RDS"))
  roc.pr.str <- read_rds(file.path(out.path, "MEIRLOP_roc_pr_str.RDS"))
  roc.pr.unstr <- read_rds(file.path(out.path, "MEIRLOP_roc_pr_unstr.RDS"))
  info.df <- read_rds(file.path(out.path, "MEIRLOP_info.RDS"))
  topMotifs.df <- read_rds(file.path(out.path, "MEIRLOP_topMotifs_lambert.RDS"))
}
```


The number of studies MEIRLOP succesfully run for: &emsp; `r length(meirlop.txts)` \

The number of studies MEIRLOP included the studied transcription factor in their results at all: &emsp; `r nrow(meirlop.list[[1]])`


## MEIRLOP rank analysis grouped by type {.tabset}

```{r meirlop_rank_analysis1, results='asis'}
meirlop.list <- rankBoxStudyType(meirlop.list, "MEIRLOP")
```

## MEIRLOP rank analysis grouped by transcription factor {.tabset}

```{r meirlop_rank_analysis2, results='asis'}
meirlop.list <- rankBoxTfFamily(meirlop.list, "MEIRLOP")
```

## MEIRLOP rank analysis grouped by number of replicates {.tabset}

```{r meirlop_rank_analysis3, results='asis'}
meirlop.list <- rankBoxReplicates(meirlop.list, "MEIRLOP")
```

## MEIRLOP rank analysis grouped by antibody type {.tabset}

```{r meirlop_rank_analysis4, results='asis'}
meirlop.list <- rankBoxAntibodyType(meirlop.list, "MEIRLOP")
```

## MEIRLOP rank analysis grouped by reference genome {.tabset}

```{r meirlop_rank_analysis5, results='asis'}
meirlop.list <- rankBoxRefGenome(meirlop.list, "MEIRLOP")
```

## MEIRLOP rank analysis grouped by bio sample {.tabset}

```{r meirlop_rank_analysis6, results='asis'}
meirlop.list <- rankBoxBioSample(meirlop.list, "MEIRLOP")
```

## MEIRLOP ROC curve {.tabset}

```{r meirlop_roc, results='asis'}
plotRoc(roc.pr.str, roc.pr.unstr, "MEIRLOP")
```

## MEIRLOP PR curve {.tabset}

```{r meirlop_pr, results='asis'}
plotPr(roc.pr.str, roc.pr.unstr, "MEIRLOP")
```

```{r saveData}
if(!file.exists(file.path(out.path, "MEIRLOP_benchmark.RDS"))) {
  # Save results as *.rds
  saveRDS(meirlop.list, file=file.path(out.path, "MEIRLOP_benchmark.RDS"))
  saveRDS(top.ten, file=file.path(out.path, "MEIRLOP_topTen.RDS"))
  saveRDS(roc.pr.str, file=file.path(out.path, "MEIRLOP_roc_pr_str.RDS"))
  saveRDS(roc.pr.unstr, file=file.path(out.path, "MEIRLOP_roc_pr_unstr.RDS"))
  saveRDS(info.df, file=file.path(out.path, "MEIRLOP_info.RDS"))
  saveRDS(topMotifs.df, file=file.path(out.path, "MEIRLOP_topMotifs_lambert.RDS"))
}
```
