---
title: "ATAC-seq CRCmapper benchmark"
author: "Leonor Schubert Santana"
date: "`r Sys.Date()`" 
site: workflowr::wflow_site
output:
   workflowr::wflow_html:
    toc: false
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=T, warning=F, message=F)
```

# Introduction

Underneath are some summary plots shown for the output of CRCmapper for all data sets in the benchmark. CRCmapper was run for each experiment in each data set using background peaks calculated by a differential peak analysis. For the plots underneath we calculated for each experiment on which rank CRCmapper finds the transcription factor examined in the study and then plot the distribution of all the ranks grouped by different variables of interest.

```{r data_preparation}
# Load libraries
library(tidyverse)
library(cowplot)
library(ggsci)

# Load utils.functions file
source("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/code/R/utils_functions.R")

# Define output path
out.path <- '/da/dmp/cb/steinse5/projects/tf_tool_benchmark/output/benchmark_data/atacseq/crcmapper/'

# Read in all knownResults.txt output files from crc for every dataset
crc.txts <- list.files("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/data/atacseq/",
                         recursive = T, full.names = T, pattern = '_DELTA_METRICS.txt')

# Create dataframe with dataset, experiment and motif name and their p-value
readIncrcTxt <- function(crc.txt) {
  conds <- unlist(str_split(crc.txt, pattern = '/'))
  crc.results <- read_tsv(crc.txt, show_col_types = FALSE) %>%
    dplyr::rename(stringent = tf) %>% 
    mutate(study = conds[11], label = conds[16], possible_ranks = length(stringent)) %>%
    filter(!is.na(stringent)) 
 
  crc.results
} 

# Check if ranking results *.rds  exists
if(!file.exists(file.path(out.path, "CRCmapper_benchmark.RDS"))) {
  
  # Apply both function to the all experiments
  crc.results <- lapply(crc.txts, readIncrcTxt)  %>% bind_rows %>%
    mutate(possible_ranks = max(possible_ranks))
  
  crc.list <- list("in_degree"=crc.results, "in_degree_norm"=crc.results, 
                   "out_degree"=crc.results, "out_degree_norm"=crc.results, 
                   "betweenness"=crc.results, "betweenness_norm"=crc.results, 
                   "alpha_centrality"=crc.results, "eigenvector"=crc.results,
                   "total_degree"=crc.results, "total_degree_norm"=crc.results)
  
  # Create empty topTen list and ROC/PR list for unstringent and stringent and info
  top.ten <- sapply(names(crc.list),function(x) NULL)
  roc.pr.str <- sapply(names(crc.list),function(x) NULL)
  roc.pr.unstr <- sapply(names(crc.list),function(x) NULL)
  info.df <- sapply(names(crc.list),function(x) NULL)
  
  # Create rankings 
  for(crc.name in names(crc.list)) {
    # Calculate for each experiment on what rank crc finds the transcription factor examined in the study
    crc.list[[crc.name]] <- crc.list[[crc.name]] %>%
      dplyr::select(c("stringent", "study", "label", "possible_ranks", all_of(crc.name))) %>%
      mutate(label = gsub('ATAC_', '', label),
             label = gsub("\\..*", "", label),
             tf = gsub('_.*', '', label),
             possible_ranks=max(possible_ranks)) %>%
      unnest()
    
    l <- createSummaryDF(crc.list[[crc.name]], crc.name, desc=TRUE, 
                         lambert=FALSE)
    
    crc.list[[crc.name]] <- l$df
    top.ten[[crc.name]] <- l$top.ten
    roc.pr.str[[crc.name]] <- l$roc.pr.str
    roc.pr.unstr[[crc.name]] <- l$roc.pr.unstr
    info.df[[crc.name]] <- l$info.df
  }
  
  
} else {
  # Read results
  crc.list <- read_rds(file.path(out.path, "CRCmapper_benchmark.RDS"))
  top.ten <- read_rds(file.path(out.path, "CRCmapper_topTen.RDS"))
  roc.pr.str <- read_rds(file.path(out.path, "CRCmapper_roc_pr_str.RDS"))
  roc.pr.unstr <- read_rds(file.path(out.path, "CRCmapper_roc_pr_unstr.RDS"))
  info.df <- read_rds(file.path(out.path, "CRCmapper_info.RDS"))
}

```

The number of studies CRCmapper succesfully run for: &emsp; `r length(crc.txts)` \

The number of studies CRCmapper included the studied transcription factor in their results at all: &emsp; `r nrow(crc.list[["in_degree"]])`

## CRCmapper rank analysis grouped by type {.tabset}

```{r crc_rank_analysis1, results='asis'}
crc.list <- rankBoxStudyType(crc.list, "CRCmapper")
```

## CRCmapper rank analysis grouped by transcription factor {.tabset}

```{r crc_rank_analysis2, results='asis'}
crc.list <- rankBoxTfFamily(crc.list, "CRCmapper")
```

## CRCmapper rank analysis grouped by number of replicates {.tabset}

```{r crc_rank_analysis3, results='asis'}
crc.list <- rankBoxReplicates(crc.list, "CRCmapper")
```

## CRCmapper rank analysis grouped by antibody type {.tabset}

```{r crc_rank_analysis4, results='asis'}
crc.list <- rankBoxAntibodyType(crc.list, "CRCmapper")
```

## CRCmapper rank analysis grouped by reference genome {.tabset}

```{r bart_rank_analysis5, results='asis'}
crc.list <- rankBoxRefGenome(crc.list, "CRCmapper")
```

## CRCmapper rank analysis grouped by bio sample {.tabset}

```{r bart_rank_analysis6, results='asis'}
crc.list <- rankBoxBioSample(crc.list, "CRCmapper")
```

## CRCmapper ROC curve {.tabset}

```{r crc_roc, results='asis'}
plotRoc(roc.pr.str, roc.pr.unstr, "CRCmapper")
```

## CRCmapper PR curve {.tabset}

```{r crc_pr, results='asis'}
plotPr(roc.pr.str, roc.pr.unstr, "CRCmapper")
```

```{r saveData}
if(!file.exists(file.path(out.path, "CRCmapper_benchmark.RDS"))) {
  # Save results as *.rds
  saveRDS(crc.list, file=file.path(out.path, "CRCmapper_benchmark.RDS"))
  saveRDS(top.ten, file=file.path(out.path, "CRCmapper_topTen.RDS"))
  saveRDS(roc.pr.str, file=file.path(out.path, "CRCmapper_roc_pr_str.RDS"))
  saveRDS(roc.pr.unstr, file=file.path(out.path, "CRCmapper_roc_pr_unstr.RDS"))
  saveRDS(info.df, file=file.path(out.path, "CRCmapper_info.RDS"))
}

#saveRDS(crc.list[["total_degree"]], file="code/R/benchmark_data/CRCmapper_benchmark.RDS")
```
