---
title: "ATAC/ChIP-seq TF tool result comparison"
author: "Leonor Schubert Santana"
date: "`r Sys.Date()`" 
site: workflowr::wflow_site
output:
   workflowr::wflow_html:
    toc: false
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=T, warning=F, message=F)
```

```{r data_preparation}
# Load libraries
library(tidyverse)
library(cowplot)
library(ggsci)

# Load utils.functions file
source("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/code/R/utils_functions.R")
source("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/code/R/read_functions.R")

# Define output path
out.path <- '/da/dmp/cb/steinse5/projects/tf_tool_benchmark/output/benchmark_data/'
fig.path <- "/da/dmp/cb/steinse5/projects/tf_tool_benchmark/output/paperPlots_revisions/"


# Find benchmark.rds 
benchmark.rds <- list.files(out.path, recursive = T, pattern = 'benchmark', full.names = T)
benchmark.rds <- benchmark.rds[-grep(benchmark.rds, pattern = '((neutral|rmd)_bg|metrics)')]
tools <- basename(benchmark.rds) %>% gsub('_.*', '', .) %>% unique()
```

## ATAC-seq hitter barplot {.tabset}

```{r atac_hitter_bar, results='asis'}
# Iterate over tools and read ATAC-seq TF ranks
ranks.df <- lapply(tools, function(tool) {
  # Subset *.rds and read label ranks
  atac.ranks <- read_rds(benchmark.rds[grep(benchmark.rds,  
                                            pattern = sprintf('atac.*%s', tool))])
  if(tool == 'CRCmapper') atac.ranks <- atac.ranks$betweenness
  
  # Reformat ATAC-seq ranks
  if(is.null(nrow(atac.ranks))) {
    ranks.df <- lapply(names(atac.ranks), function(name) {
      atac.ranks[[name]] %>%
        mutate(tool = sprintf('%s\n%s', tool, name))
    }) %>% bind_rows() 
  } else {
      ranks.df <- atac.ranks %>% mutate(tool = tool)
  }
  return(ranks.df)
}) %>% bind_rows() %>% dplyr::rename(rankStr = rank)


# Iterate over tools and read ATAC-seq TF ranks
chip.ranks <- lapply(tools, function(tool) {
  # Subset *.rds and read label ranks
  chip.ranks <- read_rds(benchmark.rds[grep(benchmark.rds,  
                                            pattern = sprintf('chip.*%s', tool))])
  #if(tool == 'CRCmapper') atac.ranks <- atac.ranks$betweenness
  
  # Reformat ChIP-seq ranks
  if(is.null(nrow(chip.ranks))) {
    ranks.df <- lapply(names(chip.ranks), function(name) {
      chip.ranks[[name]] %>%
        mutate(tool = sprintf('%s\n%s', tool, name))
    }) %>% bind_rows() 
  } else {
      ranks.df <- chip.ranks %>% mutate(tool = tool) 
  }
  return(ranks.df)
}) %>% bind_rows() %>% dplyr::rename(rankStr = rank) %>%
  filter(study %in% ranks.df$study & label %in% ranks.df$label)

# Ranks list
ranks.list <- list('ATAC-seq' = ranks.df, 'ChIP-seq' = chip.ranks)

# Prepare tool dict to tidy labels
tool.dict <- c('BART', 'CRCmapper', 'GimmeMotifs', 'HOMER + bg', 'HOMER + Lambert + bg', 'HOMER',
  'HOMER + Lambert', 'LOLA', 'MEIRLOP', 'monaLisa', 'RcisTarget + bg', 'RcisTarget',
  'TFEA')
names(tool.dict) <- ranks.df$tool %>% unique() %>% sort

# TOOLS sorted
tools.sorted <- list('stringent' = c('RcisTarget + bg', 'RcisTarget', 'monaLisa',
                                     'MEIRLOP', 'HOMER', 'HOMER + bg', 'BART',
                                     'LOLA', 'HOMER + Lambert', 'HOMER + Lambert + bg',
                                     'GimmeMotifs', 'CRCmapper', 'TFEA'),
                     'relaxed' = c('monaLisa', 'RcisTarget + bg', 'MEIRLOP', 'RcisTarget',
                                   'HOMER', 'HOMER + Lambert', 'HOMER + bg', 'GimmeMotifs',
                                   'LOLA', 'HOMER + Lambert + bg', 'BART', 'CRCmapper', 'TFEA'))

# Rank dict
rank.dict <- c('stringent' = 'rankStr', 'relaxed' = 'rankUnstr')
plot.list <- list()

for(meth in names(ranks.list)) {
  for(rn in names(rank.dict)) {
   # Compute number of hitters 
    hitter.df <-  ranks.list[[meth]] %>% 
      mutate(tool = tool.dict[tool]) %>%
      dplyr::rename(rank = as.character(rank.dict[rn])) %>%
      group_by(tool) %>%
      mutate(rank = 1+(possible_ranks*rank), 
             rank_cat = case_when(rank <= 5 ~ '5',
                                      rank > 5 & rank <= 10 ~ '10',
                                      rank > 10 & rank <= 30 ~ '30',
                                      rank > 30 ~ 'Labels recovered',
                                      is.na(rank) ~ 'Processed datasets')) %>%
      group_by(tool, rank_cat) %>%
      summarise(n = n()) 
    
    # SORT tools for plotting
    #tools.sorted <- hitter.df %>%
    #  filter(rank_cat %in% c('5', '10', '30')) %>%
    #  summarise(n = sum(n)) %>%
    #  arrange(desc(n)) 
    #tools.sorted <- 
    #  hitter.df %>% filter(!(tool %in% tools.sorted$tool)) %>% 
    #  summarise(n = sum(n)) %>% 
    #  bind_rows(tools.sorted, .)
    
    # PLOT: Top X barplot
    gg.topHittesBar <-  hitter.df  %>%
      filter(rank_cat %in% c('5', '10', '30', 'Labels recovered')) %>%
      mutate(rank_cat = factor(rank_cat, levels = rev(c( '5', '10', '30', 
                                                     'Labels recovered'))),
             tool = factor(tool, levels = tools.sorted[[rn]])) %>% #$tool)) %>%
      ggplot(aes(x = tool, y =  n, fill = rank_cat)) +
      geom_bar(stat="identity", color="black") + theme_cowplot() +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
      xlab("Method") + ylab("Count") + labs(fill=sprintf("%s\nThreshold (%s)", meth, rn)) +
      scale_fill_manual(values = c(pal_igv("default")("51")[1:4], "gray46", "white"),
                        breaks=c('Labels recovered', '30', '10', '5', 
                                 'Processed datasets', 'NA'))   
  
    plot.list[[sprintf('%s %s', meth, rn)]] <- gg.topHittesBar
    print(gg.topHittesBar)
    cat('\n\n\n')
  }
}


# Change font size
plot.list <- lapply(plot.list, function(p) {
  p + theme_cowplot(font_size = 7)  +
        theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) 
})

# Print plot
paper <- plot_grid(plotlist = plot.list, ncol=2, nrow=2,
                  align = "hv", labels = c("A", "B", "C", "D"), 
                  label_x = -0.02, axis = "tblr")

# Save plot
ggsave(paste0(fig.path, "Supp_fig_6.pdf"),
       paper, device = "pdf", width = 18, height = 12, units = "cm")


#ranks.df %>% 
#  mutate(atac_rank = 1+(rankUnstr*possible_ranks)) %>%
#  dplyr::select(tool, 1:2, atac_rank) %>%
#  left_join(., chip.ranks) %>% 
# # filter(grepl(tool, pattern = 'mona')) %>% 
#  mutate(chip_rank = 1+(rankUnstr*possible_ranks)) %>% 
#  dplyr::select(1:4, chip_rank) %>%
#  mutate(cat = case_when(atac_rank <= 30 & chip_rank <= 30 ~ 'both',
#                         atac_rank <= 30 & chip_rank > 30 ~ 'atac',
#                         atac_rank > 30 & chip_rank <= 30 ~ 'chip',
#                         atac_rank > 30 & chip_rank > 30 ~ '-')) %>%
#  group_by(tool, cat) %>% 
#  summarise(n = n())  %>%
#  filter(cat %in% c('atac', 'chip', 'both')) %>% 
#  DT::datatable()
```

## TF label rank comparison {.tabset}

```{r comparison_tf_rank, results='asis'}
# Iterate over tools
ranks.df <- lapply(tools, function(tool) {
  # Subset *.rds and read label ranks
  chip.ranks <- read_rds(benchmark.rds[grep(benchmark.rds,
                                            pattern = sprintf('chip.*%s', tool))])
  # Reformat ChIP-seq ranks
  if(is.null(nrow(chip.ranks))) {
    ranks.df <- lapply(names(chip.ranks), function(name) {
      chip.ranks[[name]][,1:3] %>%
        mutate(tool = sprintf('%s\n%s', tool, name)) %>%
        dplyr::rename(chip_rank = rank)
    }) %>% bind_rows 
    
  } else {
      ranks.df <- chip.ranks[,1:3] %>%
        mutate(tool = tool) %>%
        dplyr::rename(chip_rank = rank)
  }
  
  # Subset *.rds and read label ranks
  atac.ranks <- read_rds(benchmark.rds[grep(benchmark.rds,  
                                            pattern = sprintf('atac.*%s', tool))])
  if(tool == 'CRCmapper') atac.ranks <- atac.ranks$betweenness
  
  # Reformat ATAC-seq ranks
  if(is.null(nrow(atac.ranks))) {
    ranks.df <- lapply(names(atac.ranks), function(name) {
      atac.ranks[[name]][,1:3] %>%
        mutate(tool = sprintf('%s\n%s', tool, name)) %>%
        dplyr::rename(atac_rank = rank)
    }) %>% bind_rows() %>% left_join(ranks.df, .)
  } else {
      ranks.df <- atac.ranks[,1:3] %>%
        mutate(tool = tool) %>%
        dplyr::rename(atac_rank = rank) %>%
        left_join(ranks.df, .)
  }
  return(ranks.df)
}) %>% bind_rows()


# Tidy ranks/tools names
ranks.df <- ranks.df %>% 
  filter(grepl(tool, pattern = 'BART|gimmeMotifs|homer|lola|MEIRLOPA|monaLisa|RcisTarget|TFEA|CRCmapper')) %>%
  mutate(tool = case_when(grepl(tool, pattern = 'BART') ~ 'BART', 
                          grepl(tool, pattern = 'CRC') ~ 'CRCmapper',
                          grepl(tool, pattern = 'gimmeMotifs') ~ 'gimmeMotifs',
                          grepl(tool, pattern = 'homer\nusing no.*peaks$') ~ 'HOMER',
                          grepl(tool, pattern = 'homer\nusing no.*motifs$') ~ 'HOMER + Lambert',
                          grepl(tool, pattern = 'homer\nusing back.*peaks$') ~ 'HOMER + bg',
                          grepl(tool, pattern = 'homer\nusing back.*motifs$') ~ 'HOMER + Lambert + bg',
                          grepl(tool, pattern = 'RcisTarget.*with ') ~ 'RcisTarget + bg',
                          grepl(tool, pattern = 'RcisTarget.*without') ~ 'RcisTarget',
                          grepl(tool, pattern = 'lola') ~ 'LOLA',
                          grepl(tool, pattern = 'TFEA') ~ 'TFEA',
                          grepl(tool, pattern = 'MEIRLOP') ~ 'MEIRLOP',
                          grepl(tool, pattern = 'monaLisa') ~ 'monaLisa'))

## PLOT: Spearman correlation
#ranks.df %>%
#  group_by(tool) %>%
#  summarise(cor = list(broom::tidy(cor.test(chip_rank, atac_rank, method = 'spearman')))) %>% 
#  unnest %>% 
#  ggplot(aes(x = reorder(tool, estimate),  y = estimate)) + geom_bar(stat = 'identity') +
#  coord_flip() + theme_cowplot() + xlab('') + ylab('Spearman cor')

# PLOT: Rank scatterplot
ranks.df %>% 
  ggplot(aes(x = chip_rank, y = atac_rank)) + geom_point() +
  facet_wrap(~tool) + ggpubr::stat_cor(method = 'spearman') +
  geom_abline(linetype = 'dashed') + theme_cowplot() +
  xlab('H3K27ac TF label rank') + ylab('ATAC-seq TF label rank') 
  #theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))
```


## Jaccard index comparison of top 5/10/30 TFs {.tabset}

```{r comparison_jacc_index, results='asis'}
## Define ATAC-seq path 
#atac.path <- '/da/dmp/cb/steinse5/projects/tf_tool_benchmark/data/atacseq/'
#
## Tool dict
#tools.dict <- list('BART' = c('diffPeak_bart_results.txt', readInbartTxt),
#                   'CRCmapper' = c('DELTA_METRICS', readIncrcTxt),
#                   'gimmeMotifs' = c('final.out.txt', readInGimmeTxt),
#                   'homer_with_background' = c('knownResults.txt', readInHomerTxt),
#                   'homer_without_background' = c('knownResults.txt', readInHomerTxt),
#                   'homer_with_background_and_lambert_unstrigent' = c('knownResults.txt', readInHomerTxt),
#                   'homer_without_background_and_lambert_unstrigent' = c('knownResults.txt', readInHomerTxt),
#                   'lola' = c('lola.txt', readInLolaTxt),
#                   'MEIRLOP' = c('lr_results.tsv', readInMeirlopTxt),
#                   'monaLisa' = c('monaLisa_results.txt', readInMonalisaTxt),
#                   'RcisTarget' = c('rcistarget_without_background.txt', readInrcistargetTxt),
#                   'RcisTarget + bg' = c('rcistarget_with_background.txt', readInrcistargetTxt),
#                   'TFEA' = c('results.txt', readInTfeaTxt))
#
#
## Iterate over tools
#jaccIndex.df <- lapply(names(tools.dict), function(tool) {
#  # Find rankings 
#  ranks.txts <- list.files(atac.path, full.names = T, recursive = T,  
#                           pattern = tools.dict[[tool]][[1]], ignore.case = T)
#  ranks.txts <- ranks.txts[grep(ranks.txts, pattern = gsub(' .*', '', tool), ignore.case = T)]
#  
#  # Iterate over *.txts
#  lapply(ranks.txts, function(ranks.txt) {
#    #print(ranks.txt)
#    # Read ATAC rankings
#    atac.ranks <- tools.dict[[tool]][[2]](ranks.txt)
#    # Read ChIP-seq rankings
#    chip.txt <- gsub('atacseq', 'chipseq', ranks.txt) %>% 
#      gsub('ATAC', 'H3K27ac',  .)
#    if(!file.exists(chip.txt)) return(NULL)
#    chip.ranks  <- tools.dict[[tool]][[2]](chip.txt)
#    # Compute JaccIndex on top 5/10/30 ranks
#    jis <- sapply(c(5, 10, 30), function(ii) {
#      atac.ttfs <- atac.ranks$motif_name[1:ii]
#      chip.ttfs <- chip.ranks$motif_name[1:ii]
#      sum(atac.ttfs %in% chip.ttfs)/length(unique(c(atac.ttfs, chip.ttfs)))
#      
#    })
#    tibble(tool = tool, study = atac.ranks$study[1],
#           cond = atac.ranks$cond[1],
#           top = c(5, 10, 30), jacc_index = jis)
#  }) %>% bind_rows()
#})
#
#
## PLOT: Jaccard index
#bind_rows(jaccIndex.df) %>% 
#  mutate(top = as.factor(top),
#         tool = ifelse(tool == 'lola', 'LOLA', tool),
#         tool = ifelse(tool == 'homer_without_background', 'HOMER', tool),
#         tool = ifelse(tool == 'homer_with_background', 'HOMER + bg', tool),
#         tool = ifelse(tool == 'homer_without_background_and_lambert_unstrigent', 'HOMER + Lambert', tool),
#         tool = ifelse(tool == 'homer_with_background_and_lambert_unstrigent', 'HOMER + Lambert + bg', tool)) %>%
#  ggplot(aes(x = reorder(tool, jacc_index), y = jacc_index, fill = top)) + geom_boxplot()  +
#  xlab('') + ylab('Jaccard index') + coord_flip() +
#  scale_fill_viridis_d(name = 'Threshold') +
#  theme_cowplot()
#
```

