---
title: "ChIP-seq - GimmeMotifs benchmark"
author: "Leonor Schubert Santana"
date: "`r Sys.Date()`" 
site: workflowr::wflow_site
output:
   workflowr::wflow_html:
    toc: false
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=T, warning=F, message=F)
```

# Introduction

Underneath are some summary plots shown for the output of gimmeMotifs for all data sets in the benchmark. GimmeMotifs was run for each experiment in each data set using background peaks calculated by a differential peak analysis. For the plots underneath we calculated for each experiment on which rank gimmeMotifs finds the transcription factor examined in the study and then plot the distribution of all the ranks grouped by different variables of interest.

```{r data_preparation}
# Load libraries
library(tidyverse)
library(cowplot)
library(jsonlite)
library(ggsci)

# Load utils.functions file
source("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/code/R/utils_functions.R")

# Define output path
out.path <- '/da/dmp/cb/steinse5/projects/tf_tool_benchmark/output/benchmark_data/chipseq/gimmemotifs/'

# Read in all knownResults.txt output files from gimme for every dataset
gimme.txts <- list.files("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/data/chipseq/",
                          recursive = T, full.names = T, pattern = 'final.out.txt')

# Create dataframe with dataset, experiment and motif name and their p-value
readInGimmeTxt <- function(gimme.txt) {
  conds <- unlist(str_split(gimme.txt, pattern = '/'))
  gimme.results <- read_tsv(gimme.txt, show_col_types = FALSE) 
  columnNames <- colnames(gimme.results)[1:3]
  
  if (grepl(columnNames[2], pattern = '(KD|KO|antagonist)')){
    scoreIndex <- 3
  } else if (grepl(columnNames[3], pattern = '(KD|KO|antagonist)')){
    scoreIndex <- 2
  } else {
    scoreIndex <-which(grepl(columnNames, pattern = '(OE|agonist)'))
  }
  
  gimme.results <- gimme.results %>%
    dplyr::rename(z_score=columnNames[scoreIndex], motif_name="...1") %>%
    mutate(study = conds[11], label = conds[16], possible_ranks=length(motif_name)) %>%
    dplyr::select(study, label, motif_name, z_score, possible_ranks)
 
  gimme.results
} 

# Check if ranking results *.rds  exists
if(!file.exists(file.path(out.path, "gimmeMotifs_benchmark.RDS"))) {
  # Apply both function to the all experiments
  gimme.results <- lapply(gimme.txts, readInGimmeTxt)  %>% bind_rows
  
  gimme.list <- list('using no background peaks' = gimme.results)
  
  # Create empty topTen list and ROC/PR list for unstringent and stringent and info
  top.ten <- sapply(names(gimme.list),function(x) NULL)
  roc.pr.str <- sapply(names(gimme.list),function(x) NULL)
  roc.pr.unstr <- sapply(names(gimme.list),function(x) NULL)
  info.df <- sapply(names(gimme.list),function(x) NULL)
  topMotifs.df <- sapply(names(gimme.list),function(x) NULL)
  
  
  # Create rankings 
  for(gimme.name in names(gimme.list)) {
    l <- createSummaryDF(gimme.list[[gimme.name]], "z_score", desc=TRUE, 
                         lambert=TRUE)
    
    gimme.list[[gimme.name]] <- l$df
    top.ten[[gimme.name]] <- l$top.ten
    roc.pr.str[[gimme.name]] <- l$roc.pr.str
    roc.pr.unstr[[gimme.name]] <- l$roc.pr.unstr
    info.df[[gimme.name]] <- l$info.df
    topMotifs.df[[gimme.name]] <- l$topMotifs.df
  }
} else {
  gimme.list <- readRDS(file.path(out.path, "gimmeMotifs_benchmark.RDS"))
  top.ten <- readRDS(file.path(out.path, "gimmeMotifs_topTen.RDS"))
  roc.pr.str <- readRDS(file.path(out.path, "gimmeMotifs_roc_pr_str.RDS"))
  roc.pr.unstr <- readRDS(file.path(out.path, "gimmeMotifs_roc_pr_unstr.RDS"))
  info.df <- readRDS(file.path(out.path, "gimmeMotifs_info.RDS"))
  topMotifs.df <- readRDS(file.path(out.path, "gimmeMotifs_topMotifs_lambert.RDS"))
}

```

The number of studies gimmeMotifs succesfully run for: &emsp; `r length(gimme.txts)` \

The number of studies gimmeMotifs included the studied transcription factor in their results at all: &emsp; `r nrow(gimme.list[[1]])`


## GimmeMotifs rank analysis grouped by type {.tabset}

```{r gimme_rank_analysis1, results='asis'}
gimme.list <- rankBoxStudyType(gimme.list, "GimmeMotifs")
```

## GimmeMotifs rank analysis grouped by transcription factor {.tabset}

```{r gimme_rank_analysis2, results='asis'}
gimme.list <- rankBoxTfFamily(gimme.list, "GimmeMotifs")
```

## GimmeMotifs rank analysis grouped by number of replicates {.tabset}

```{r gimme_rank_analysis3, results='asis'}
gimme.list <- rankBoxReplicates(gimme.list, "GimmeMotifs")
```

## GimmeMotifs rank analysis grouped by antibody type {.tabset}

```{r gimme_rank_analysis4, results='asis'}
gimme.list <- rankBoxAntibodyType(gimme.list, "GimmeMotifs")
```

## GimmeMotifs rank analysis grouped by reference genome {.tabset}

```{r gimme_rank_analysis5, results='asis'}
gimme.list <- rankBoxRefGenome(gimme.list, "GimmeMotifs")
```

## GimmeMotifs rank analysis grouped by bio sample {.tabset}

```{r gimme_rank_analysis6, results='asis'}
gimme.list <- rankBoxBioSample(gimme.list, "GimmeMotifs")
```

## GimmeMotifs ROC curve {.tabset}

```{r gimme_roc, results='asis'}
plotRoc(roc.pr.str, roc.pr.unstr, "GimmeMotifs")
```

## GimmeMotifs PR curve {.tabset}

```{r gimme_pr, results='asis'}
plotPr(roc.pr.str, roc.pr.unstr, "GimmeMotifs")
```

```{r saveData}
if(!file.exists(file.path(out.path, "gimmeMotifs_benchmark.RDS"))) {
  saveRDS(gimme.list, file=file.path(out.path, "gimmeMotifs_benchmark.RDS"))
  saveRDS(top.ten, file=file.path(out.path, "gimmeMotifs_topTen.RDS"))
  saveRDS(roc.pr.str, file=file.path(out.path, "gimmeMotifs_roc_pr_str.RDS"))
  saveRDS(roc.pr.unstr, file=file.path(out.path, "gimmeMotifs_roc_pr_unstr.RDS"))
  saveRDS(info.df, file=file.path(out.path, "gimmeMotifs_info.RDS"))
  saveRDS(topMotifs.df, file=file.path(out.path, "gimmeMotifs_topMotifs_lambert.RDS"))
}
```
