---
title: "ChIP-seq - RcisTarget benchmark"
author: "Leonor Schubert Santana"
date: "`r Sys.Date()`" 
site: workflowr::wflow_site
output:
   workflowr::wflow_html:
    toc: false
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=T, warning=F, message=F)
```

# Introduction

Underneath are some summary plots shown for the output of RcisTarget for all data sets in the benchmark. RcisTarget was run for each experiment in each data set. For the plots underneath we calculated for each experiment on which rank RcisTarget finds the transcription factor examined in the study and then plot the distribution of all the ranks grouped by different variables of interest.

```{r data_preparation}
# Load libraries
library(tidyverse)
library(cowplot)
library(ggsci)

# Load utils.functions file
source("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/code/R/utils_functions.R")

# Define output path
out.path <- '/da/dmp/cb/steinse5/projects/tf_tool_benchmark/output/benchmark_data/chipseq/rcistarget/'

# Read in all rcistarget output files for every experiment
rcistarget.txts <- list.files("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/data/chipseq/",
                          recursive = T, full.names = T, pattern = 'rcistarget')
rcistarget.txts <- rcistarget.txts[-grep(rcistarget.txts, pattern = 'benchmark_(rmd|neutral)_bg')]
rcistarget.nb <- rcistarget.txts[grep(rcistarget.txts, pattern = 'without_background')]
rcistarget.b <- rcistarget.txts[grep(rcistarget.txts, pattern = 'with_background')]


# Create dataframe with dataset, experiment and motif name and their p-value
readInrcistargetTxt <- function(rcistarget.txt) {
  conds <- unlist(str_split(rcistarget.txt, pattern = '/'))
  rcistarget.results <- read_delim(rcistarget.txt, delim="\t") %>%
  dplyr::rename(motif_name = 2, stringent = 5, nes = 3) %>%
  mutate(study = conds[11], 
         label = conds[16]) %>%
  dplyr::select(study, label, motif_name, stringent, nes)
  rcistarget.results <- rcistarget.results %>%
    mutate(possible_ranks=nrow(rcistarget.results)) 
  
  rcistarget.results
} 

# Check if ranking results *.rds  exists
if(!file.exists(file.path(out.path, "RcisTarget_benchmark.RDS"))) {
  
  # Apply function to all the rcistarget output files of all experiments
  rcistarget.resultsNB <- lapply(rcistarget.nb, readInrcistargetTxt)  %>% 
    bind_rows %>%
    mutate(possible_ranks=max(possible_ranks))
  rcistarget.resultsB <- lapply(rcistarget.b, readInrcistargetTxt)  %>% 
    bind_rows %>%
    mutate(possible_ranks=max(possible_ranks))
  
  rcistarget.list <- list('with background'=rcistarget.resultsB,
                          'without background'=rcistarget.resultsNB)
  
  # Create empty topTen list and ROC/PR list for unstringent and stringent and info
  top.ten <- sapply(names(rcistarget.list),function(x) NULL)
  roc.pr.str <- sapply(names(rcistarget.list),function(x) NULL)
  roc.pr.unstr <- sapply(names(rcistarget.list),function(x) NULL)
  info.df <- sapply(names(rcistarget.list),function(x) NULL)
  
  # Create rankings 
  for(rcistarget.name in names(rcistarget.list)) {
    # Calculate for each experiment on what rank rcistarget finds the transcription factor examined in the study
    rcistarget.list[[rcistarget.name]] <- rcistarget.list[[rcistarget.name]] %>%
      mutate(stringent = gsub("\\s*\\([^\\)]+\\)", "", stringent),
             stringent = gsub("\\.", ";", stringent),
             stringent = gsub("; $", "", stringent),
             stringent = gsub("NANA$", "", stringent),
             # unstringent = gsub("\\s*\\([^\\)]+\\)", "", unstringent),
             # unstringent = gsub("\\.", ";", unstringent),
             # unstringent = gsub("; $", "", unstringent),
             # unstringent = gsub("NANA$", "", unstringent),
             # unstringent = case_when(!is.na(stringent) ~ paste(stringent, unstringent, sep="; "),
             #                         is.na(stringent) ~ unstringent),
             label = gsub('H3K27(A|a)c_', '', label),
             label = gsub("\\..*", "", label),
             tf = gsub('_.*', '', label)) %>%
      filter(!is.na(stringent))
    
    l <- createSummaryDF(rcistarget.list[[rcistarget.name]],
                         "nes", desc=TRUE, lambert=FALSE)
    
    rcistarget.list[[rcistarget.name]] <- l$df
    top.ten[[rcistarget.name]] <- l$top.ten
    roc.pr.str[[rcistarget.name]] <- l$roc.pr.str
    roc.pr.unstr[[rcistarget.name]] <- l$roc.pr.unstr
    info.df[[rcistarget.name]] <- l$info.df
  }
  
 
} else {
  # Read results
  rcistarget.list <- read_rds(file.path(out.path, "RcisTarget_benchmark.RDS"))
  top.ten <- read_rds(file.path(out.path, "RcisTarget_topTen.RDS"))
  roc.pr.str <- read_rds(file.path(out.path, "RcisTarget_roc_pr_str.RDS"))
  roc.pr.unstr <- read_rds(file.path(out.path, "RcisTarget_roc_pr_unstr.RDS"))
  info.df <- read_rds(file.path(out.path, "RcisTarget_info.RDS"))
}
```

The number of studies RcisTarget without using background succesfully run for: &emsp; `r length(rcistarget.nb)` \

The number of studies RcisTarget without using background included the studied transcription factor in their results at all: &emsp; `r nrow(rcistarget.list[[1]])`

The number of studies RcisTarget with using background succesfully run for: &emsp; `r length(rcistarget.b)` \

The number of studies RcisTarget with using background included the studied transcription factor in their results at all: &emsp; `r nrow(rcistarget.list[[2]])`

## RcisTarget rank analysis grouped by type {.tabset}

```{r rcistarget_rank_analysis1, results='asis'}
rcistarget.list <- rankBoxStudyType(rcistarget.list, "rcisTarget")
```

## RcisTarget rank analysis grouped by transcription factor {.tabset}

```{r rcistarget_rank_analysis2, results='asis'}
rcistarget.list <- rankBoxTfFamily(rcistarget.list, "rcisTarget")
```

## RcisTarget rank analysis grouped by number of replicates {.tabset}

```{r rcistarget_rank_analysis3, results='asis'}
rcistarget.list <- rankBoxReplicates(rcistarget.list, "rcisTarget")
```

## RcisTarget rank analysis grouped by antibody type {.tabset}

```{r rcistarget_rank_analysis4, results='asis'}
rcistarget.list <- rankBoxAntibodyType(rcistarget.list, "rcisTarget")
```

## RcisTarget rank analysis grouped by reference genome {.tabset}

```{r rcistarget_rank_analysis5, results='asis'}
rcistarget.list <- rankBoxRefGenome(rcistarget.list, "rcisTarget")
```

## RcisTarget rank analysis grouped by bio sample {.tabset}

```{r rcistarget_rank_analysis6, results='asis'}
rcistarget.list <- rankBoxBioSample(rcistarget.list, "RcisTarget")
```

## RcisTarget ROC curve {.tabset}

```{r rcistarget_roc, results='asis'}
plotRoc(roc.pr.str, roc.pr.unstr, "RcisTarget")
```

## RcisTarget PR curve {.tabset}

```{r rcistarget_pr, results='asis'}
plotPr(roc.pr.str, roc.pr.unstr, "RcisTarget")
```

```{r saveData}
if(!file.exists(file.path(out.path, "RcisTarget_benchmark.RDS"))) {
  # Save results as *.rds
  saveRDS(rcistarget.list, file=file.path(out.path, "RcisTarget_benchmark.RDS"))
  saveRDS(top.ten, file=file.path(out.path, "RcisTarget_topTen.RDS"))
  saveRDS(roc.pr.str, file=file.path(out.path, "RcisTarget_roc_pr_str.RDS"))
  saveRDS(roc.pr.unstr, file=file.path(out.path, "/RcisTarget_roc_pr_unstr.RDS"))
  saveRDS(info.df, file=file.path(out.path, "RcisTarget_info.RDS"))
}
```

