---
title: "ChIP-seq Homer benchmark (rmd bg)"
author: "Leonor Schubert Santana"
date: "`r Sys.Date()`" 
site: workflowr::wflow_site
output:
   workflowr::wflow_html:
    toc: false
    code_folding: hide
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=T, warning=F, message=F)
```

# Introduction

Underneath are some summary plots shown for the output of HOMER for all data sets in the benchmark. HOMER was run for each experiment in each data set once using background peaks calculated by a differential peak analysis and once letting HOMER calculate the background distribution. For the plots underneath we calculated for each experiment on which rank HOMER finds the transcription factor examined in the study and then plot the distribution of all the ranks grouped by different variables of interest.

```{r data_preparation}
# Load libraries
library(tidyverse)
library(cowplot)
library(jsonlite)
library(ggsci)

# Load utils.functions file
source("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/code/R/utils_functions.R")

# Define output path
out.path <- '/da/dmp/cb/steinse5/projects/tf_tool_benchmark/output/benchmark_data/chipseq_rmd_bg/homer/'
if(!dir.exists(out.path)) dir.create(out.path, recursive = T)

# Read in all knownResults.txt output files from homer for every dataset
homer.txts <- list.files("/da/dmp/cb/steinse5/projects/tf_tool_benchmark/data/chipseq/",
                          recursive = T, full.names = T, pattern = 'knownResults.txt')
homer.txts <- homer.txts[grep(homer.txts, pattern = 'benchmark_rmd_bg')]
homer.nL <- homer.txts[-grep(homer.txts, pattern = 'lambert')]
homer.L <- homer.txts[grep(homer.txts, pattern = 'lambert_unstrigent')]
homer.b <- homer.nL[grep(homer.nL, pattern = 'with_background')]
homer.bL <- homer.L[grep(homer.L, pattern = 'with_background')]

# Create dataframe with dataset, experiment and motif name and their p-value
readInHomerTxt <- function(homer.txt) {
  conds <- unlist(str_split(homer.txt, pattern = '/'))
  homer.results <- read_tsv(homer.txt, show_col_types = FALSE) %>%
  dplyr::rename(motif_name = 1, p_value = 3) %>%
  mutate(study = conds[11], label = conds[16], possible_ranks=length(motif_name)) %>%
  dplyr::select(study, label, motif_name, p_value, possible_ranks)
 
  homer.results
} 

# Check if ranking results *.rds  exists
if(!file.exists(file.path(out.path, "homer_rmd_bg_benchmark.RDS"))) {
  # Apply both function to the all experiments
  homer.resultsB <- lapply(homer.b, readInHomerTxt)  %>% bind_rows
  homer.resultsBL <- lapply(homer.bL, readInHomerTxt)  %>% bind_rows
  homer.list <- list('using background peaks' = homer.resultsB, 
                     'using background peaks and lambert motifs' = homer.resultsBL)
  
  # Create empty topTen list and ROC/PR list for unstringent and stringent and info
  top.ten <- sapply(names(homer.list),function(x) NULL)
  roc.pr.str <- sapply(names(homer.list),function(x) NULL)
  roc.pr.unstr <- sapply(names(homer.list),function(x) NULL)
  info.df <- sapply(names(homer.list),function(x) NULL)
  topMotifs.df <- sapply(names(homer.list),function(x) NULL)
  
  
  # Create rankings 
  for(homer.name in names(homer.list)) {
    if (grepl(homer.name, pattern="lambert")) {
      l <- createSummaryDF(homer.list[[homer.name]], "p_value", desc=FALSE, 
                           lambert=TRUE)
      
      homer.list[[homer.name]] <- l$df
      top.ten[[homer.name]] <- l$top.ten
      roc.pr.str[[homer.name]] <- l$roc.pr.str
      roc.pr.unstr[[homer.name]] <- l$roc.pr.unstr
      info.df[[homer.name]] <- l$info.df
      topMotifs.df[[homer.name]] <- l$topMotifs.df
      
    } else {
      # Calculate for each experiment on what rank homer finds the transcription factor examined in the study
      homer.list[[homer.name]] <- homer.list[[homer.name]] %>%
        mutate(stringent = gsub("\\(.*", "", motif_name),
               stringent = gsub("/.*", "", stringent),
               label = gsub('H3K27(A|a)c_', '', label),
               label = gsub("\\..*", "", label),
               possible_ranks=nrow(read_delim(homer.bL[1], delim="\t")),
               tf = gsub('_.*', '', label)) %>%
        unnest() %>%
        filter(!is.na(stringent))
      
      l <- createSummaryDF(homer.list[[homer.name]], "p_value", desc=FALSE, 
                           lambert=FALSE)
      
      homer.list[[homer.name]] <- l$df
      top.ten[[homer.name]] <- l$top.ten
      roc.pr.str[[homer.name]] <- l$roc.pr.str
      roc.pr.unstr[[homer.name]] <- l$roc.pr.unstr
      info.df[[homer.name]] <- l$info.df
    }
  }

  
} else {
  homer.list <- file.path(out.path, "homer_rmd_bg_benchmark.RDS")
  top.ten <- file.path(out.path, "homer_rmd_bg_topTen.RDS")
  roc.pr.str <- file.path(out.path, "homer_rmd_bg_roc_pr_str.RDS")
  roc.pr.unstr <- file.path(out.path, "homer_rmd_bg_roc_pr_unstr.RDS")
  info.df <- file.path(out.path, "homer_rmd_bg_info.RDS")
  topMotifs.df <- file.path(out.path, "homer_rmd_bg_topMotifs_lambert.RDS")
}


```

The number of studies Homer with background succesfully run for: &emsp; `r length(homer.b)` \

The number of studies Homer with background included the studied transcription factor in their results at all: &emsp; `r nrow(homer.list[[1]])`

The number of studies Homer with background but using Lambert succesfully run for: &emsp; `r length(homer.bL)` \

The number of studies Homer with background but using Lambert included the studied transcription factor in their results at all: &emsp; `r nrow(homer.list[[2]])`


## homer rank analysis grouped by type {.tabset}

```{r homer_rank_analysis1, results='asis'}
homer.list <- rankBoxStudyType(homer.list, "homer")
```

## homer rank analysis grouped by transcription factor {.tabset}

```{r homer_rank_analysis2, results='asis'}
homer.list <- rankBoxTfFamily(homer.list, "homer")
```

## homer rank analysis grouped by number of replicates {.tabset}

```{r homer_rank_analysis3, results='asis'}
homer.list <- rankBoxReplicates(homer.list, "homer")
```

## homer rank analysis grouped by antibody type {.tabset}

```{r homer_rank_analysis4, results='asis'}
homer.list <- rankBoxAntibodyType(homer.list, "homer")
```

## homer rank analysis grouped by reference genome {.tabset}

```{r homer_rank_analysis5, results='asis'}
homer.list <- rankBoxRefGenome(homer.list, "homer")
```

## homer rank analysis grouped by bio sample {.tabset}

```{r homer_rank_analysis6, results='asis'}
homer.list <- rankBoxBioSample(homer.list, "homer")
```

## homer ROC curve {.tabset}

```{r gimme_roc, results='asis'}
plotRoc(roc.pr.str, roc.pr.unstr, "homer")
```

## homer PR curve {.tabset}

```{r gimme_pr, results='asis'}
plotPr(roc.pr.str, roc.pr.unstr, "homer")
```

```{r saveData}
if(!file.exists(file.path(out.path, "homer_rmd_bg_benchmark.RDS"))) {
  # Save results as *.rds
  saveRDS(homer.list, file=file.path(out.path, "homer_rmd_bg_benchmark.RDS"))
  saveRDS(top.ten, file=file.path(out.path, "homer_rmd_bg_topTen.RDS"))
  saveRDS(roc.pr.str, file=file.path(out.path, "homer_rmd_bg_roc_pr_str.RDS"))
  saveRDS(roc.pr.unstr, file=file.path(out.path, "homer_rmd_bg_roc_pr_unstr.RDS"))
  saveRDS(info.df, file=file.path(out.path, "homer_rmd_bg_info.RDS"))
  saveRDS(topMotifs.df, file=file.path(out.path, "homer_rmd_bg_topMotifs_lambert.RDS"))
}
```
